graphql_compiler/tests/test_schema_fingerprint.py

```diff
@@ -0,0 +1,245 @@
+
+        # Assert that the computed fingerprint is not the same if we add a new field.
+        schema_text_with_added_field = """
+            type Object{
+                field1: String
+                field3: String
+                field4: String
+                field2: String
+                field5: String
+            }
+        """
+        schema_with_added_field_fingerprint = compute_schema_fingerprint(
+            build_ast_schema(parse(schema_text_with_added_field))
+        )
+        self.assertNotEqual(schema_with_added_field_fingerprint, fingerprint)
+
+    def test_interface_order(self):
+        schema_text1 = """
+            type Object implements Interface1 & Interface2 {
+                field1: String
+                field2: String
+            }
+
+            interface Interface1 {
+                field1: String
+            }
+
+            interface Interface2 {
+                field1: String
+            }
+        """
+        schema_text2 = """
+            type Object implements Interface2 & Interface1 {
+                field1: String
+                field2: String
+            }
+
+            interface Interface1 {
+                field1: String
+            }
+
+            interface Interface2 {
+                field1: String
+            }
+        """
+        _compare_schema_fingerprints(self, schema_text1, schema_text2)
+
+    def test_union_order(self):
+        schema_text1 = """
+            type Object1 {
+                field1: String
+            }
+
+            type Object2 {
+                field2: String
+            }
+
+            union UnionType = Object1 | Object2
+        """
+        schema_text2 = """
+            type Object1 {
+                field1: String
+            }
+
+            type Object2 {
+                field2: String
+            }
+
+            union UnionType = Object2 | Object1
+        """
+        _compare_schema_fingerprints(self, schema_text1, schema_text2)
+
+    def test_top_level_order(self):
+        schema_text1 = """
+            scalar Date
+
+            directive @output on FIELD
+
+            type Object {
+                field1: String
+            }
+
+            union UnionType = Object
+        """
+        schema_text2 = """
+            directive @output on FIELD
+
+            scalar Date
+
+            union UnionType = Object
+
+            type Object {
+                field1: String
+            }
+        """
+        _compare_schema_fingerprints(self, schema_text1, schema_text2)
+
+    def test_directive_definition_argument_order(self):
+        schema_text1 = """
+            directive @filter(
+                op_name: String!
+                value: [String!]
+            ) repeatable on FIELD | INLINE_FRAGMENT
+        """
+        schema_text2 = """
+            directive @filter(
+                value: [String!]
+                op_name: String!
+            ) repeatable on FIELD | INLINE_FRAGMENT
```

Comment:
What happens if we have a repeatable directive on `FIELD_DEFINITION` and it is applied twice to the same field? Is the lexicographic sorting function going to order those two occurrences of the directive in a specific way?

Presumably, the following:
```graphql
type Foo {
    my_field: Int @custom_directive(b: "456", a: 1) @custom_directive(b: "123", a: 23)
}
```
after lexicographic sorting should turn into
```graphql
type Foo {
    my_field: Int @custom_directive(a: 23, b: "123") @custom_directive(a: 1, b: "456")
}
```
and it would be good to verify that schemas like that have the same fingerprints with a test.

Separately, it might be a good idea to check if the GraphQL-core library has tests like this for the lexicographic ordering function, and contribute them in a PR if not.

Reply:
I decided to remove this test entirely. I realized that the GraphQLSchema object does not capture information about directive located at field definitions, (other than for the deprecated directive). In other words, the GraphQLField class does not have a directives field or anything of the sort.

This made me realize though that I was missing tests where the fingerprint was different. (I would have realized that this sooner if I had added a test that expected that two fingerprints were different based on having different directives on field definitions). I added a number of this style of tests.

Reply:
Ignoring directives applied to schema elements such as fields and types is not a robust plan. We cannot promise to calculate semantically meaningful fingerprints over "most" of the data for an ill-defined "most".

At the very least, add the tests you removed back in, but rewrite them in such a way that they point out the problematic behavior and fail, then mark them `xfail` to signal that they are known issues and open an issue on Github to resolve them.

We may also be able to use something like the approach in https://github.com/graphql-python/graphql-core-next/issues/65 to get directive values for types, fields, and other schema elements, and add them into the hash separately.

Hopefully https://github.com/graphql/graphql-js/issues/1789 gets addressed before too long (and perhaps we should help it along if you feel like learning some Javascript) so we can do this more thoroughly via the lexicographic sort function.